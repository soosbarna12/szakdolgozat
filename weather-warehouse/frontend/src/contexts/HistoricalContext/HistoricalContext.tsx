import React, { createContext, useEffect, useState } from "react";
import { LOCAL_STORAGE_TEMPERATURE_SCALE } from "../../consts/temperatureScale.const";
import { TemperatureScale } from "../../types/temperatureScale.type";
import { HistoricalContextType, HistoricalLocationData, HistoricalPageData, HistoricalProviderProps } from "./HistoricalContext.type";


export const HistoricalContext = createContext<HistoricalContextType>({
  location: { name: "", lat: 0, lon: 0 },
  setLocation: () => { },
  historicalPageData: [] as any,
  setHistoricalPageData: () => { },
  temperatureScale: TemperatureScale.Celsius,
  setTemperatureScale: () => { },
});

export const HistoricalLocationProvider = ({ children }: HistoricalProviderProps) => {

  const [location, setLocation] = useState<HistoricalLocationData>(() => {
    const storedLocation = localStorage.getItem("historicalLocation");
    const defaultLocation = { name: "", lat: 0, lon: 0 };
    try {
      return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
    } catch (error) {
      console.error("Failed to parse location from localStorage:", error);
      return defaultLocation;
    }
  });

  const [historicalPageData, setHistoricalPageData] = useState<HistoricalPageData[]>(() => {
    const storedData = localStorage.getItem("historicalPageData");
    try {
      return storedData ? JSON.parse(storedData) : [];
    } catch (error) {
      console.error("Failed to parse historical data from localStorage:", error);
      return [];
    }
  });

  const [temperatureScale, setTemperatureScale] = useState<TemperatureScale | string>(() => {
    const storedTemperatureScale = localStorage.getItem(LOCAL_STORAGE_TEMPERATURE_SCALE);
    try {
      return storedTemperatureScale ? JSON.parse(storedTemperatureScale) : TemperatureScale.Celsius;
    } catch (error) {
      console.error("Failed to parse temperature scale from localStorage:", error);
      return TemperatureScale.Celsius;
    }
  });


  useEffect(() => {
    if (historicalPageData) {
      localStorage.setItem("historicalPageData", JSON.stringify(historicalPageData));
    }
  }, [historicalPageData]);

  useEffect(() => {
    localStorage.setItem("historicalLocation", JSON.stringify(location));
  }, [location]);

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_TEMPERATURE_SCALE, JSON.stringify(temperatureScale));
  }, [temperatureScale]);


  return (
    <HistoricalContext.Provider value={{ location, setLocation, historicalPageData, setHistoricalPageData, temperatureScale, setTemperatureScale }}>
      {children}
    </HistoricalContext.Provider>
  );
};
