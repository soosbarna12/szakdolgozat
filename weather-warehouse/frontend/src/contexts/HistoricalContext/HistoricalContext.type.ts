import { Dispatch, ReactNode, SetStateAction } from "react";
import { HistoricalDataTable } from "../../types/historicalDataTable.type";
import { TemperatureScale } from "../../types/temperatureScale.type";

export interface HistoricalLocationData {
  name: string;
  country?: string;
  lat: number;
  lon: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface HistoricalPageData extends HistoricalDataTable {}

export interface HistoricalContextType {
  location: HistoricalLocationData;
  historicalPageData: any [];
  temperatureScale: TemperatureScale | string;
  setLocation: Dispatch<SetStateAction<HistoricalLocationData>>;
  setHistoricalPageData: Dispatch<SetStateAction<HistoricalPageData[]>>;
  setTemperatureScale: Dispatch<SetStateAction<TemperatureScale | string>>;
}

export interface HistoricalProviderProps {
  children: ReactNode;
}
