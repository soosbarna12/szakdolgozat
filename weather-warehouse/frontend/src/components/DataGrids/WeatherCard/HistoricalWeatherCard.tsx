import { Box, Skeleton, Typography, useTheme } from "@mui/material";
import dayjs from "dayjs";
import React, { useContext } from "react";
import { HistoricalContext } from "../../../contexts/HistoricalContext/HistoricalContext";
import { TemperatureScale } from "../../../types/temperatureScale.type";
import { temperatureScaleChanger } from "../../../utils/temperatureScaleChanger";
import { HistoricalWeatherCardProps } from "./HistoricalWeatherCard.type";


export function HistoricalWeatherCard({ data }: Readonly<HistoricalWeatherCardProps>) {

  const theme = useTheme();
  const { temperatureScale } = useContext(HistoricalContext);
  const temperatureScaleLabel = (temperatureScale === TemperatureScale.Celsius ? "°C" : "°F");

  if (!data) {
    return (
      <Skeleton
        data-testid="historicalWeatherCardSkeleton"
        variant="rectangular"
        animation="wave"
        width="100%"
        height="100%"
        sx={{ borderRadius: '10px' }}
      />
    );
  }

  function getFormattedTime(date: string) {
    return dayjs(date).format("dddd, MMMM D, YYYY");
  };

  return (
    <Box sx={{ textAlign: "left" }}>
      <Typography data-testid="date" variant="h6">{getFormattedTime(data.date)}</Typography>
      <Typography data-testid="location" variant="h5" fontWeight="700">
        {data.cityName}, {data.countryCode}
      </Typography>
      <Box sx={{ display: "flex", alignItems: "center", m: 0, p: 0 }} color={theme.palette.primary.dark}>

        <Typography data-testid="temperature" variant="h2" sx={{ ml: 1, m: 0 }}>
          {temperatureScaleChanger(TemperatureScale.Celsius, temperatureScale as TemperatureScale, data.temp)}
          <span style={{ fontSize: "0.5em", verticalAlign: "super" }}>{temperatureScaleLabel}</span>
        </Typography>
      </Box>
      <Box>
        <Typography data-testid="precipitation">
          Precipitation: {data.precipitation || 0} mm/h
        </Typography>
        <Typography data-testid="pressure">Pressure: {data.pressure} hPa</Typography>
        <Typography data-testid="windSpeed">Wind: {data.windSpeed} m/s</Typography>
      </Box>
    </Box>
  );
}
