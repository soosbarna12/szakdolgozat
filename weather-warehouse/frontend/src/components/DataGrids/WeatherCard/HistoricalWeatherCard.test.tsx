import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import React from 'react';
import { HistoricalContext } from '../../../contexts/HistoricalContext/HistoricalContext';
import { TemperatureScale } from '../../../types/temperatureScale.type';
import { renderWithQueryClient } from '../../../utils/test/renderWithQueryClient';
import { HistoricalWeatherCard } from './HistoricalWeatherCard';

describe('DataGrids/WeatherCard/HistoricalWeatherCard', () => {
    const mockData = {
        date: '2023-01-01',
        cityName: 'New York',
        countryCode: 'US',
        temp: 5,
        precipitation: 10,
        pressure: 1015,
        windSpeed: 5,
    };

    const renderComponent = (data = mockData, temperatureScale = TemperatureScale.Celsius) => {
        return renderWithQueryClient(
            <HistoricalContext.Provider value={{ temperatureScale } as any}>
                <HistoricalWeatherCard data={data as any} />
            </HistoricalContext.Provider>
        );
    };

    it('matches HistoricalWeatherCard component snapshot', () => {
        const component = renderComponent();
        expect(component).toMatchSnapshot();
    });

    it('renders skeleton when data is not provided', () => {
        renderComponent(null);
        const skeleton = screen.getByTestId('historicalWeatherCardSkeleton');
        expect(skeleton).toBeInTheDocument();
    });

    it('renders the correct date, city, and country', () => {
        renderComponent();
        expect(screen.getByTestId('date')).toHaveTextContent('Sunday, January 1, 2023');
        expect(screen.getByTestId('location')).toHaveTextContent('New York, US');
    });

    it('renders temperature with the correct scale and label', () => {
        renderComponent();
        const temperature = screen.getByTestId('temperature');
        expect(temperature).toHaveTextContent('5°C');
    });

    it('renders precipitation, pressure, and wind speed', () => {
        renderComponent();
        expect(screen.getByTestId('precipitation')).toHaveTextContent('Precipitation: 10 mm/h');
        expect(screen.getByTestId('pressure')).toHaveTextContent('Pressure: 1015 hPa');
        expect(screen.getByTestId('windSpeed')).toHaveTextContent('Wind: 5 m/s');
    });

    it('renders temperature in Fahrenheit when scale is set to Fahrenheit', () => {
        renderComponent(mockData, TemperatureScale.Fahrenheit);
        const temperature = screen.getByTestId('temperature');
        expect(temperature).toHaveTextContent('41°F');
    });
});