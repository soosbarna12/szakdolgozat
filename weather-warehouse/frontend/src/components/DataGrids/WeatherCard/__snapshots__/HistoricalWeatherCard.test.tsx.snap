// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGrids/WeatherCard/HistoricalWeatherCard matches HistoricalWeatherCard component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-13brihr"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
          data-testid="date"
        >
          Sunday, January 1, 2023
        </h6>
        <h5
          class="MuiTypography-root MuiTypography-h5 css-stnzmu-MuiTypography-root"
          data-testid="location"
        >
          New York
          , 
          US
        </h5>
        <div
          class="MuiBox-root css-nkffzc"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 css-6fhuif-MuiTypography-root"
            data-testid="temperature"
          >
            5
            <span
              style="font-size: 0.5em; vertical-align: super;"
            >
              °C
            </span>
          </h2>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="precipitation"
          >
            Precipitation: 
            10
             mm/h
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="pressure"
          >
            Pressure: 
            1015
             hPa
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="windSpeed"
          >
            Wind: 
            5
             m/s
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-13brihr"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
        data-testid="date"
      >
        Sunday, January 1, 2023
      </h6>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-stnzmu-MuiTypography-root"
        data-testid="location"
      >
        New York
        , 
        US
      </h5>
      <div
        class="MuiBox-root css-nkffzc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-6fhuif-MuiTypography-root"
          data-testid="temperature"
        >
          5
          <span
            style="font-size: 0.5em; vertical-align: super;"
          >
            °C
          </span>
        </h2>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          data-testid="precipitation"
        >
          Precipitation: 
          10
           mm/h
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          data-testid="pressure"
        >
          Pressure: 
          1015
           hPa
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          data-testid="windSpeed"
        >
          Wind: 
          5
           m/s
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
